internal/models/openai/openai_embedding.go

```diff
@@ -0,0 +1,238 @@
+func (eb *ByIndex) Swap(i, j int) {
+	eb.resp.Data[i], eb.resp.Data[j] = eb.resp.Data[j], eb.resp.Data[i]
+}
+func (eb *ByIndex) Less(i, j int) bool { return eb.resp.Data[i].Index < eb.resp.Data[j].Index }
+
+type ErrorInfo struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+	Param   string `json:"param,omitempty"`
+	Type    string `json:"type"`
+}
+
+type EmbedddingError struct {
+	Error ErrorInfo `json:"error"`
+}
+
+type OpenAIEmbeddingInterface interface {
+	Check() error
+	Embedding(modelName string, texts []string, dim int, user string, timeoutSec time.Duration) (*EmbeddingResponse, error)
+}
+
+type openAIBase struct {
+	apiKey string
+	url    string
+}
+
+func (c *openAIBase) Check() error {
+	if c.apiKey == "" {
+		return fmt.Errorf("api key is empty")
+	}
+
+	if c.url == "" {
+		return fmt.Errorf("url is empty")
+	}
+	return nil
+}
+
+func (c *openAIBase) genReq(modelName string, texts []string, dim int, user string) *EmbeddingRequest {
+	var r EmbeddingRequest
+	r.Model = modelName
+	r.Input = texts
+	r.EncodingFormat = "float"
+	if user != "" {
+		r.User = user
+	}
+	if dim != 0 {
+		r.Dimensions = dim
+	}
+	return &r
+}
+
+type OpenAIEmbeddingClient struct {
+	openAIBase
+}
+
+func NewOpenAIEmbeddingClient(apiKey string, url string) *OpenAIEmbeddingClient {
+	return &OpenAIEmbeddingClient{
+		openAIBase{
+			apiKey: apiKey,
+			url:    url,
+		},
+	}
+}
+
+func (c *OpenAIEmbeddingClient) Embedding(modelName string, texts []string, dim int, user string, timeoutSec time.Duration) (*EmbeddingResponse, error) {
+	r := c.genReq(modelName, texts, dim, user)
+	data, err := json.Marshal(r)
+	if err != nil {
+		return nil, err
+	}
+
+	if timeoutSec <= 0 {
+		timeoutSec = 30
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutSec*time.Second)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.url, bytes.NewBuffer(data))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.apiKey))
+	body, err := utils.RetrySend(req, 3)
+	if err != nil {
+		return nil, err
+	}
+	var res EmbeddingResponse
+	err = json.Unmarshal(body, &res)
+	if err != nil {
+		return nil, err
+	}
+	sort.Sort(&ByIndex{&res})
+	return &res, err
+}
+
+type AzureOpenAIEmbeddingClient struct {
+	openAIBase
+	apiVersion string
+}
+
+func NewAzureOpenAIEmbeddingClient(apiKey string, url string) *AzureOpenAIEmbeddingClient {
+	return &AzureOpenAIEmbeddingClient{
+		openAIBase: openAIBase{
+			apiKey: apiKey,
+			url:    url,
+		},
+		apiVersion: "2024-06-01",
+	}
+}
+
+func (c *AzureOpenAIEmbeddingClient) Embedding(modelName string, texts []string, dim int, user string, timeoutSec time.Duration) (*EmbeddingResponse, error) {
+	r := c.genReq(modelName, texts, dim, user)
+	data, err := json.Marshal(r)
+	if err != nil {
+		return nil, err
+	}
+
+	if timeoutSec <= 0 {
```

Comment:
consider make default timeoutSec a constant in common.go? I see this number and default retry times 3 used in several places.